File: TKTT_HK4/2.0. Tìm Kiếm/DFS/DFS.cpp
#include <bits/stdc++.h>
using namespace std;

/*
input: so luong dinh, sl canh: 9 9

*/

int n, m;              // luu sl canh, dinh
vector<int> adj[1001]; // mang cac vector luu ds ke
bool visited[1001];    // mang kiem tra da duyet chua
queue<int> s;

void input()
{
    // chuyen danh sach ke
    cin >> n >> m;
    for (int i = 0; i < m; i++)
    {
        int x, y;
        cin >> x >> y;
        adj[x].push_back(y);
        adj[y].push_back(x); // do thi co huong thi khong pushback dong nay
    }
    memset(visited, false, sizeof(visited));
}

void DFS(int u)
{
    s.push(u);
    cout << u << " ";
    visited[u] = true; // danh dau la dc tham
    // xet cac dinh ke cua u
    for (auto v : adj[u])
    {
        if (!visited[v]) // neu chua tham thi de quy den tham
        {
            DFS(v);
            s.push(u);
        }
    }
}

int main()
{
    input();
    DFS(1);
    cout << "\nPath: ";
    while (!s.empty()) // In ra đường đi trong queue
    {
        cout << s.front() << " ";
        s.pop();
    }
    return 0;
}
/*
9 9
1 2
1 3
1 5
2 4
3 6
3 7
3 9
5 8
8 9
*/

File: TTUD_HK5/ThucHanh/TH3/DFS.py
def dfs(u):
    stack = [u]
    while stack:
        v = stack.pop()
        if not visited[v]:
            print(v, end=" ")
            visited[v] = True
            for x in adj[v]:
                if not visited[x]:
                    stack.append(x)


if __name__ == "__main__":
    n = 7
    visited = [False] * n
    adj = [
        [1, 2, 3],
        [3, 4],
        [5],
        [2, 5, 6],
        [3, 6],
        [],
        [5],
    ]

    dfs(0)
